### PR TITLE
move `async unsafe fn` into `#![feature(async_unsafe)]`

### LINKED ISSUE #62500: move async unsafe fn to distinct feature gate
Per @Centril's [comment](https://github.com/rust-lang/rust/issues/62149#issuecomment-508947381), we want to move `async unsafe fn` to its own feature gate:

> I would also like to move `async unsafe fn` out of the MVP into its own feature gate because I think a) it has seen little use, b) it is not particularly well tested, c) it ostensibly behaves weirdly because the `.await` point is not where you write `unsafe { ... }` and this is understandable from "leaky implementation POV" but not so much from an effects POV, d) it has seen little discussion and was not included in the RFC nor this report, and e) we did this with `const fn` and it worked fine. (I can write up the feature gating PR)

This issue has been assigned to @delan via [this comment](https://github.com/rust-lang/rust/issues/62500#issuecomment-509460671).

### PR DESCRIPTION
**Pull request by delan**
This is my first PR to the project, so please let me know if anything’s amiss!

- tracking issue #62501 
- closes #62500 

r? @Centril

### CODE REVIEW

File: src/test/ui/async-await/async-await.rs [hunk 1]
```diff
@@ -2,21 +2,23 @@
...
 use std::sync::{
     Arc,
     atomic::{self, AtomicUsize},
 };
 use std::task::{Context, Poll};
 use arc_wake::ArcWake;
+use wake_once::wake_and_yield_once;
 
-struct Counter {
+pub struct Counter {
```
Comment (delan) [inline src/test/ui/async-await/async-await.rs]: spurious change made when I tried to extract Counter into an aux-build

Comment (delan) [reply src/test/ui/async-await/async-await.rs]: fixed in 4ea974281ab22967a707df4b1575a9e35a871962

**Comment by Centril** [comment]
@delan  Thank you for the PR! ❤️ 

After some discussion, I was convinced not to feature gate this after all (see https://github.com/rust-lang/rust/issues/62149#issuecomment-510541229) so I don't think we can merge this PR. I hope this hasn't soured your taste for contributing to `rustc` and hopefully you got to learn about how to contribute to the compiler.

One thing that is however sorely lacking is tests for `async unsafe fn`. Would you like to file another PR for that so that we:
- Check that an `async unsafe fn` requires `unsafe { ... }` to be called.
   - Including `async unsafe fn` implementations...
- Tangentially also test the behavior of `unsafe { async { $stuff } }` wrt. allowing unsafe operations in `$stuff`.

If you would like, [there are some other tests needed](https://github.com/rust-lang/rust/issues/62121#issuecomment-506884048) that we could use some help with. :)

Thanks again for the PR, and sorry about the noise!

**Comment by delan** [comment]
> After some discussion, I was convinced not to feature gate this after all (see [#62149 (comment)](https://github.com/rust-lang/rust/issues/62149#issuecomment-510541229)) so I don't think we can merge this PR. I hope this hasn't soured your taste for contributing to `rustc` and hopefully you got to learn about how to contribute to the compiler.

Not at all, these things happen!

> One thing that is however sorely lacking is tests for `async unsafe fn`. Would you like to file another PR for that so that we:
> 
>     * Check that an `async unsafe fn` requires `unsafe { ... }` to be called.
>       
>       * Including `async unsafe fn` implementations...
> 
>     * Tangentially also test the behavior of `unsafe { async { $stuff } }` wrt. allowing unsafe operations in `$stuff`.

I’ll see what I can do about those later today.

> Thanks again for the PR, and sorry about the noise!

Happy to help!

**Pull request closed without merging**